opnode/rollup/derive/payload_attributes.go

```diff
@@ -176,45 +180,151 @@ func UserDeposits(height uint64, receipts []*types.Receipt) ([]*types.DepositTx,
+	Time() uint64
+}
+
+// PayloadAttributes derives a sequence of pre-execution payload attributes from:
+//  - The L1 information the L1-info deposit is derived from
+//  - The L1 receipts the user deposits are derived from
+//  - The data batches the L2 sequencer work is derived from
+//  - The L2 information of the block the new derived blocks build on
+//
 // This is a pure function.
-func PayloadAttributes(block BlockInput, receipts []*types.Receipt) (*l2.PayloadAttributes, error) {
-	if !CheckReceipts(block, receipts) {
-		return nil, fmt.Errorf("receipts are not consistent with the block's receipts root: %s", block.ReceiptHash())
+func PayloadAttributes(config *rollup.Config, l1Info L1Info, receipts []*types.Receipt, seqWindow []BatchData, l2Info L2Info) ([]*l2.PayloadAttributes, error) {
+	// Retrieve the deposits of this epoch (all deposits from the first block)
+	deposits, err := DeriveDeposits(l1Info, receipts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive deposits: %v", err)
 	}
 
-	l1Tx := types.NewTx(L1InfoDeposit(block))
+	// copy L1 randomness (mix-digest becomes randao field post-merge)
+	// TODO: we don't have a randomness oracle on L2, what should sequencing randomness look like.
+	// Repeating the latest randomness of L1 might not be ideal.
+	randomnessSeed := l2.Bytes32(l1Info.MixDigest())
+
+	// Collect all L2 batches, the batches may be out-of-order, or possibly missing.
+	l2Blocks := make(map[uint64]*l2.PayloadAttributes)
+	highestSeenTimestamp := l1Info.Time()
+	for _, batch := range seqWindow {
+
+		// Track the last batch we've seen (gaps will be filled with empty L2 blocks)
+		if batch.Timestamp > highestSeenTimestamp {
+			highestSeenTimestamp = batch.Timestamp
+		}
+
+		l2Blocks[batch.Timestamp] = &l2.PayloadAttributes{
+			Timestamp:             l2.Uint64Quantity(batch.Timestamp),
+			Random:                randomnessSeed,
+			SuggestedFeeRecipient: config.FeeRecipientAddress,
+			Transactions:          batch.Transactions,
+		}
+	}
+
+	// If there are no submitted batches, at least derive the deposit block.
+	if len(seqWindow) == 0 {
+		highestSeenTimestamp += config.BlockTime
+	}
+
+	// fill the gaps and always ensure at least one L2 block
+	// TODO: What happens if `highestSeenTimestamp` start rolling into the next epoch (higher than the next L1 block timestamp)
+	var out []*l2.PayloadAttributes
+	for t := l1Info.Time() + config.BlockTime; t <= highestSeenTimestamp; t += config.BlockTime {
+		if bl, ok := l2Blocks[t]; ok {
+			out = append(out, bl)
+		} else {
+			// skipped/missing L2 block, create an empty block instead
+			out = append(out, &l2.PayloadAttributes{
+				Timestamp:             l2.Uint64Quantity(t),
+				Random:                randomnessSeed,
+				SuggestedFeeRecipient: config.FeeRecipientAddress,
+				Transactions:          nil,
+			})
+		}
+	}
+
+	// Force deposits into the first block
+	out[0].Transactions = append(append(make([]l2.Data, 0), deposits...), out[0].Transactions...)
```

Comment:
TODO: when inserting the user deposited transactions, they need to be placed between the system deposit tx and the batch transactions

Reply:
Done, but a little hacky.